lldp: T7165: add support to enable only rx/tx on specific interfaces

c-po opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->

LLDP is a stateless protocol which does not necessitate sending to receive advertisements. There are multiple scenarios such as provider peering links in which it is advantageous to receive LLDP but not disclose internal information to the provider.

Add new CLI commands:
`set service lldp interface <name> mode [disable|rx-and-tx|rx-only|tx-only]`

The default is unchanged and will be rx-and-tx.

Furthermore if an interface has an explicit LLDP disable configured under `set service lldp interface <name> disable` this will be migrated to `set service lldp interface <name> mode disable`


## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
* https://vyos.dev/T7165

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
* https://github.com/vyos/vyos-documentation/pull/1601

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_lldp.py
test_01_lldp_basic (__main__.TestServiceLLDP.test_01_lldp_basic) ... ok
test_02_lldp_mgmt_address (__main__.TestServiceLLDP.test_02_lldp_mgmt_address) ... ok
test_03_lldp_interfaces (__main__.TestServiceLLDP.test_03_lldp_interfaces) ... ok
test_04_lldp_all_interfaces (__main__.TestServiceLLDP.test_04_lldp_all_interfaces) ... ok
test_05_lldp_location (__main__.TestServiceLLDP.test_05_lldp_location) ... ok
test_06_lldp_snmp (__main__.TestServiceLLDP.test_06_lldp_snmp) ... ok
test_07_lldp_interface_mode (__main__.TestServiceLLDP.test_07_lldp_interface_mode) ... ok

----------------------------------------------------------------------
Ran 7 tests in 13.116s

OK
```

![image](https://github.com/user-attachments/assets/72d515fb-b841-4b6d-9930-4287035ace53)


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

dmbaturin commented on interface-definitions/service_lldp.xml.in:
```diff
@@ -29,7 +29,34 @@
+                <properties>
+                  <help>Set LLDP receive/transmit operation mode of this interface</help>
+                  <completionHelp>
+                    <list>disable rx-and-tx tx-only rx-only</list>
```
I wonder if just `tx`, `rx`, and `tx-rx` would suffice.

c-po replied:
I opted for rx-tx to keep the alphabetical order.

dmbaturin commented on interface-definitions/service_lldp.xml.in:
```diff
@@ -29,7 +29,34 @@
+                  </completionHelp>
+                  <valueHelp>
+                    <format>disable</format>
+                    <description>Disable both receive/transmit</description>
```
```suggestion
                    <description>Do not process or send LLDP messages</description>
```

dmbaturin commented on interface-definitions/service_lldp.xml.in:
```diff
@@ -29,7 +29,34 @@
+                  </valueHelp>
+                  <valueHelp>
+                    <format>rx-and-tx</format>
+                    <description>Enable both receive/transmit</description>
```
```suggestion
                    <description>Send and process LLDP messages</description>
```

dmbaturin commented on interface-definitions/service_lldp.xml.in:
```diff
@@ -29,7 +29,34 @@
+                  </valueHelp>
+                  <valueHelp>
+                    <format>rx-only</format>
+                    <description>Enable only receive</description>
```
```suggestion
                    <description>Process incoming LLDP messages</description>
```

dmbaturin commented on interface-definitions/service_lldp.xml.in:
```diff
@@ -29,7 +29,34 @@
+                  </valueHelp>
+                  <valueHelp>
+                    <format>tx-only</format>
+                    <description>Enable only transmit</description>
```
```suggestion
                    <description>Send LLDP messages</description>
```

dmbaturin reviewed (changes requested):
I'd change the wording of the help strings a bit; I also wonder if option names can be made more concise. No objections against the implementation.

c-po committed:
lldp: T7165: add support to enable only rx or tx on specific interface

LLDP is a stateless protocol which does not necessitate sending to receive
advertisements. There are multiple scenarios such as provider peering links in
which it is advantageous to receive LLDP but not disclose internal information
to the provider.

Add new CLI command:
* set service lldp interface <name> mode [disable|rx-tx|rx|tx]

The default is unchanged and will be rx-tx.

Furthermore if an interface has an explicit LLDP disable configured under
"set service lldp interface <name> disable" this will be migrated to
"set service lldp interface <name> mode disable"

sempervictus commented:
Thank you @c-po
